### PR TITLE
[CBOR] Implement support for tagged types

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Contributes to #32046 and #32047.

- [x] Implement `PeekTag()` method.
- [x] Implement DateTime type support.
- [x] Implement BigInteger support.
- [x] Implement `decimal` support.

Addendum:
- [x] Add WriteInt32()/ReadInt32() convenience methods.
- [x] Add pervasive buffer postcondition checks on negative reader tests and fix relevant bugs.
- [x] Rename "Special" major type to "Simple".

### CODE REVIEW

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:57]: There's no baked-in format string for exclusively recognizing RFC3339 dates. As such, I'm using `DateTimeOffset.Parse` instead of `DateTimeOffset.ParseExact`, at the cost of tolerating certain nonstandard string formats. I'm assuming this shouldn't be a big problem given this is a reader.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:44]: This method will recognize both unix time and string datetime formats. I'm wondering whether it might be appropriate that those be exposed as separate methods.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:48]: Note that unlike the primitive reader methods, reader implementations for tagged values might partially advance the reader state if there is a format error (note that the same is also true for the `SkipValue()` method).

I'm wondering whether this might justify using some sort of checkpointing logic so that the reader state can be rolled back in case of an error, but I'm not sure whether the benefits would justify the added complexity here.

Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:48]: The only time SkipValue can fail is when the content is corrupt and moving forward again would not succeed.

It should generally be true that calling the wrong read method does not corrupt the reader (or, if you prefer, any wrong read call utterly destroys the reader).

So either this should avoid advancing at all until the end (which probably requires passing some offsets into some helpers that previously didn't need them), or the body should be something like

```C#
ReadOnlyMemory<byte> checkpoint = _buffer;

try
{
    // current code here
}
catch
{
    _buffer = checkpoint;
    throw;
}
```

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:57]: Isn't it just `"yyyy-MM-ddTHH:mm:ss.sssK"`?

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:60]: Maybe, do you think it would be worth enforcing on the reader's side?

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs]: It probably would need a bit more work than that in the general case (e.g. resetting counters, popping items off the stack, etc). Wondering whether it might be worth exposing as protected methods for folks extending the Reader with new types, i.e.
```csharp
protected SomeOpaqueCheckpointToken CreateCheckpoint();
protected RestoreCheckpoint(SomeOpaqueCheckpointToken token);
```

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs]: Assuming we're ok with state changes on format errors, it might be possible to dodge the above monstrosity and just use `PeekTag()`, at least for DateTime.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Tag.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Tag.cs:139]: This could probably use some tests for pre-1970. Especially from POSIX time (since 2.4.1 says it uses negative numbers to represent that).

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:45]: Perhaps it just needs a defaulted `CborTag? expectedTag` parameter, that currently supports `null` (don't care, just do it) and the two formats.  Other inputs would be an ArgumentOutOfRangeException until/unless new DateTimeOffset-compatibile tag types get added.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:87]: There's a loss of reader consistency on the exceptions after ReadTag()... the reader should probably rewrind.

FWIW, the fact that these methods can throw when Skip won't does reinforce for me the opinion that Skip shouldn't validate UTF-8 encodings for text string values.  (I think the real decision comes when conformance modes are added... does Skip verify that the integer `2` is not lazily encoded across multiple bytes, vs being in the compact single-byte form?  If yes, then Skip validating "primitive" values and not validating tagged values is consistent.  If no, then it should only validate what it needs to maintain state and the ability to jump to the next value).

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:104]: I can go either way here, but I have to ask:

Since the DateTimeOffset method supports multiple formats, should ReadBigInteger also support major type 0 and major type 1, since they can be upconverted losslessly?

I haven't looked at a whole lot of schema, so I don't know if any are flexible on "small integer" vs "big integer" encodings.  The only one I can really think of that could even really theoretically use numbers in that range is RSA public key encoding; but https://tools.ietf.org/html/rfc8230#section-4 defines its own encoding (Unsigned Big-Endian value with extraneous leading zeros removed encoded into a byte string).

And, of course, someone could add support for it with an extension method, like

```C#
public static ReadAnyBigInteger(this CborReader reader)
{
    switch (reader.State)
    {
        case StateEnum.NonNegativeInteger:
        case StateEnum.NegativeInteger:
            return reader.ReadInt64();
        case StateEnum.TaggedValue:
            return reader.ReadBigInteger();
    }

    throw new Exception("I'm too lazy to write good messages in hypothetical code");
}
```

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Integer.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Integer.cs:36]: I'm not sure what this even means.  `WriteCborNegativeIntegerEncoding(1)` => 0b001_0001 => `-2`?

Perhaps my question is "does this need to be public? would callers figure out how to use it correctly?"

Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Integer.cs:36]: (And if it's public, it needs tests :smile:)

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Tag.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Tag.cs:33]: The reader supports reading fractional seconds, but the writer doesn't support writing them?

I would ask "should this format also accept a DateTimeOffset value since it can do the calculations itself?", but then there's the question of "do you want to include fractional seconds, or not?".  (Or, the simple assertion that "no, you don't"; and callers who do can just manually call `WriteUnixTimeSeconds((value - DateTimeOffset.UnixEpoch).TotalSeconds)`)

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Tag.cs:33]: The reader supports fractional seconds, since they're part of the CBOR spec. I was contemplating potentially making the writer accept a double argument, however in the end I went for making it consistent with `DateTimeOffset.FromUnixTimeSeconds(long)`.

> "should this format also accept a DateTimeOffset value since it can do the calculations itself?"

Probably not, since this would result in loss of information for any value with a nonzero offset. I also considered DateTime, however I find the integer choice to be more straightforward.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Integer.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Integer.cs:36]: > I'm not sure what this even means. WriteCborNegativeIntegerEncoding(1) => 0b001_0001 => -2?

Yes. this is replicating the already existing `ReadCborNegativeIntegerEncoding()`. I wasn't convinced initially but it seems that this might be necessary for encoding certain semantic values (such as BigInteger).

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:104]: > Since the DateTimeOffset method supports multiple formats, should ReadBigInteger also support major type 0 and major type 1, since they can be upconverted losslessly?

I was working under the assumption that we were avoiding implicit conversions by design. For the case of `ReadDateTimeOffset()` I'm contemplating whether I should revert to distinct readers for each tagged type.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:87]: Agreed that adding conformance levels should add desired degree of control wrt Skip behaviour.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:45]: Hmm, I'm not sure I'm a big fan of that, it also means it introduces a new failure mode for tags that don't match either of the two expected values. I'd probably either leave as-is or break up into two separate methods:

```csharp
DateTimeOffset ReadDateTimeString(); // for lack of a better name for RFC3339 date strings
DateTimeOffset ReadUnixDateTime();
```
As you noted in a different comment, it should be trivial for consumers to recompose into one extension method that handles both cases.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Integer.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Integer.cs:36]: Added tests

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Tag.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Tag.cs:142]: Done

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs:357]: To consider: should we mark the checkpoint methods public or protected? If so these asserts should be changed to runtime checks.

**Comment by eiriktsarpalis** [comment]
@bartonjs PR is now ready for review. There are still a few design issues to be considered, in the unresolved conversations above.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Tag.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Tag.cs:136]: Since I can't read CBOR the way I can DER: Are any of these using indefinite strings for the date?  I assume there should be some that do... (if it's verboten, negative tests, otherwise, this positive test).

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Tag.cs:136]: The spec is vague in that regard. Here it is in its entirety:
```
Tag value 0 is for date/time strings that follow the standard format described in [RFC3339], as refined by Section 3.3 of [RFC4287].
```
I'd therefore assume they should be accepted, provided the conformance level for the reader permits it.

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Tag.cs:137]: Addressed.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:44]: Your proposal to split the two methods renamed this to `ReadDateTimeString`, but you left the name as `ReadDateTimeOffset`.  Given that you've split them I would recommend that you do go forward with the rename.

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:44]: I felt that keeping this name is appropriate, in the sense that rfc3339 strings faithfully represent DateTimeOffset values in a canonical way (unlike unix times). This is also reflecting the `void WriteDateTimeOffset(DateTimeOffset)` method which is used to write string datetimes.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs:357]: Making it public would require further validation, like that the checkpoint came from this instance (otherwise the stack wouldn't have the same meaning).

You'd also need to account for more complicated things, like the stack being popped and a new frame pushed (complete the current array, start the next array/map/tag, then rewind) which there's currently not a need for (and what happens if you restore a checkpoint to a different reader instance?).

So, I'd say "if a scenario warrants it, OK, but I'd hold off unless it's clearly needed; then you need to decide if you're using the public checkpoint internally, or this sort of slim checkpoint")

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs:357]: > You'd also need to account for more complicated things, like the stack being popped and a new frame pushed.

That's already accounted for by storing the buffer offset in each frame. It's currently only being used in the assert.

> So, I'd say "if a scenario warrants it, OK, but I'd hold off unless it's clearly needed; then you need to decide if you're using the public checkpoint internally, or this sort of slim checkpoint")

I feel this is already in the borderline of over-engineering, but I thought I'd get a second opinion regardless :-)

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:34]: I don't think I see any tests for PeekTag... presumably a good test case is to encode the same DateTimeOffset as both a string and the Unix epoch seconds and do the correct read just using the reader API.

Tests for unmapped (future) tag values are also good.  Like, write the code to read this future schema:

```
Tag value 0xFFFFFFFF_FFFFFFFF represents a date, optionally with a time of day, using a calendar other than the Gregorian calendar (a calendar date).

A calendar date is represented as a tagged array. The first value in the array is an integer identifying the calendar system from a table below. This is optionally followed by the English name of the calendar, a string with tag 119. This is optionally followed by the natural language name of the calendar, a string with tag 120. This is optionally followed by the name of an era (for era based calendars), a string with tag 729. This is followed by the day on the calendar.  If the calendar date represents a specific time on that date, there will be one additional element, the number of seconds indicated after the start of that date.

The date portion of a calendar date can be specified as either:
* A string, representing the date in yyyy-MM-dd syntax
* A number, representing the number of days since the beginning of the calendar era.

[for a challenge, the table of calendar IDs has been omitted]
```

Assuming I did all of my math right, an array of 4 such values is

84DBFFFFFFFFFFFFFFFF8405D9031066E698ADE5928C1A5017200019ECDBDBFFFFFFFFFFFFFFFF83026A353732392D31312D303919ECDBDBFFFFFFFFFFFFFFFF8406D878782DD9B1D984D8AAD98ED991D982D992D988D990D98AD98520D9B1D984D992D987D990D8ACD992D8B1D990D98AD9916A313338392D30352D313019ECDBDBFFFFFFFFFFFFFFFF8407D8776B556D6D20616C2D517572611A82395A801989C6

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs:34]: Out of curiosity, where did you find that schema?

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Tag.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Tag.cs:90]: @tannergooding you might be interested in reviewing this.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs]: Will revisit this issue when implementing conformance levels.

**Review by bartonjs — DISMISSED** [review]
Next up: A more comprehensive test that reads and writes COSE public keys (https://tools.ietf.org/html/rfc8152#appendix-C.7.1 has some EC examples; but examples that use the RSA extensions from https://tools.ietf.org/html/rfc8230 are also/likely-more valueable).

To consider:
* The addition of the Int32 accelerators has made for some mild asymmetry. Consider adding UInt32 accelerators as well, for symmetry.

File: src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Integer.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Integer.cs:133]: Please add the overflow and bad input variant tests.

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Integer.cs:133]: I have the changes locally, but waiting on current CI to finish before I push them.

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Integer.cs:133]: Which seems might take a while :-/

**Review by bartonjs — APPROVED** [review]

**Pull request merged**
